### PR TITLE
Support multiple DB URLs in each named database configuration; Rotation of database connections.

### PR DESCRIPTION
**Pull request by jason-instacart**
### Problems
Jardin currently supports one database URL per named-configuration. For example, for `shoppers_read`, we can only point it to one replica. 

In the deployments, however, we have multiple replicas for `shoppers_read` and we want the application to distribute reads evenly across all replicas.

### Solutions

1) Allow whitespace-separated multiple DB URLs in jardin_conf.py, for each named configuration. 

2) By default, a process just randomly picks one of the connections, and sticks to it the entire time - maintain connection stickiness in a process/session/task/job, etc.

3) Provide a way to remove stickiness to a single connection and shuffle the connections. This can be called at the end of a session/task/job in a long-running process.

### Test Performed

- Existing unit tests
- Plenty of console tests using production jardin_config.py
- Integration test in a pre-production environment

### Notes

- This PR does not have the support for multiple configs in dictionary format. It only supports multiple URLs (string format)

- The change is intended to be backward compatible. Single URL per named config should continue to work.

### CODE REVIEW

File: jardin/model.py [hunk 1]
```diff
@@ -491,13 +490,8 @@ def _default_table_name(self):
 
     @classmethod
     def db(self, role='replica', db_name=None):
-        if not hasattr(self, '_db'): self._db = {}
         name = db_name or self.db_names.get(role)
-        
-        if name not in self._db:
-            self._db[name] = DatabaseConnections.connection(name)
```
Comment (jason-instacart) [inline jardin/model.py:498]: No need to cache connection in model class. The connection object is already cached in DatabaseConnections.

File: jardin/model.py [hunk 2]
```diff
@@ -437,17 +437,16 @@ def find(self, id, **kwargs):
 
     @classmethod
     def db_adapter(self, role='replica', db_name=None):
-        if not hasattr(self, '_db_adapter'):
-            self._db_adapter = {}
+        if not hasattr(self, '_db_metadata'):
+            self._db_metadata = {}
         db_name = db_name or self.db_names.get(role)
         key = '%s_%s' % (db_name, role)
-        if key not in self._db_adapter:
```
Comment (jason-instacart) [inline jardin/model.py:444]: No need to cache DatabaseAdapter here - make it be consistent with the way we run raw SQLs (jardin.query()).

However, caching model_metadata is still needed.

File: jardin/database/__init__.py [hunk 3]
```diff
@@ -15,42 +17,64 @@ class UnsupportedDriver(Exception): pass
...
-    def urls(self, name):
-        if len(self._urls) == 0:
+    def db_configs(self, name):
+        if len(self._db_configs) == 0:
             config.init()
-        for (nme, url) in config.DATABASES.items():
-            if url:
-                self._urls[nme] = DatabaseConfig(url)
-        return self._urls[name]
+            for (nme, urls) in config.DATABASES.items():
```
Comment (sherin) [inline jardin/database/__init__.py:62]: i think you mean to have this for loop outside the if

File: jardin/config.py [hunk 4]
```diff
@@ -25,4 +26,8 @@ def init():
     logger = logging.getLogger('jardin')
     logger.setLevel(LOG_LEVEL)
     
-    setattr(this, 'INITIALIZED', True)
\ No newline at end of file
+    setattr(this, 'INITIALIZED', True)
+
+
+def reset_session():
```
Comment (sherin) [inline jardin/config.py:32]: i am not sure if you need this defined in config or did you mean to have this in jardin/__init__.py ?

File: jardin/database/__init__.py (hunk 3 above)
Comment (jason-instacart) [reply jardin/database/__init__.py:62]: Why outside? I think the original code had a bug about this. 

We only need to run the for-loop if the cache collection (old: self._urls; new: self._db_configs), is empty.

File: jardin/config.py (hunk 4 above)
Comment (jason-instacart) [reply jardin/config.py:32]: I can move it outside config

File: jardin/database/__init__.py (hunk 3 above)
Comment (sherin) [reply jardin/database/__init__.py:62]: ah got it, makes sense. reading this again.

File: jardin/database/__init__.py [hunk 5]
```diff
@@ -15,42 +17,64 @@ class UnsupportedDriver(Exception): pass
...
+                if not urls:
+                    continue
+                url_list = re.split(r'\s+', urls) if isinstance(urls, str) else [urls]
```
Comment (mathieuripert) [inline jardin/database/__init__.py:65]: if `urls` is already a `list`, should not it be `url_list = re.split(r'\s+', urls) if isinstance(urls, str) else urls`

File: jardin/__init__.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
     from jardin.model import Model, Collection
     from jardin.query import query
     from jardin.transaction import Transaction
+    from jardin.tools import reset_session
```
Comment (sherin) [inline jardin/__init__.py:7]: nice

File: jardin/database/__init__.py (hunk 5 above)
Comment (jason-instacart) [reply jardin/database/__init__.py:65]: jardin supports two formats of DB config: string url and dictionary (snowflake). In the PR, we don’t deal with multi config in dictionary format - so just put it in array to process by map(). Let me make a comment in code

Comment (mathieuripert) [reply jardin/database/__init__.py:66]: Ok i see ! 
I thought we could also define a db config this way:
`"db_name" : [replica_url_1, replica_url_2, replica_url_3]`

Comment (mathieuripert) [reply jardin/database/__init__.py:66]: But your change makes sense !

**Pull request merged**